### PR TITLE
Windows support

### PR DESCRIPTION
**Pull request by Motakjuq**
Add windows support

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/62)

### CODE REVIEW

**Comment by manishrjain** [comment]
Thanks a lot for the change. Can you please merge master so the conflict resolves. Also, got a question, if you can answer that.

---

Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[table/mmap_windows.go, line 10 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/62#-KnMW4GO-O15sw5DcO5c:-KnMW4GPJtARwe0KVPkv:b-4bq7ya) ([raw file](https://github.com/dgraph-io/badger/blob/6ade7de36a488ff5517f4ba4f9a6ff99ec7f16e0/table/mmap_windows.go#L10)):*
> ```Go
> 
> func mmap(fd *os.File, size int64) ([]byte, error) {
> 	handler, err := syscall.CreateFileMapping(syscall.Handle(fd.Fd()), nil, syscall.PAGE_READONLY, uint32(size>>32), uint32(size), nil)
> ```

Over 100 chars. Split into 2 lines.

---

*[table/mmap_windows.go, line 14 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/62#-KnMW9pY-X3wdRDDUobH:-KnMW9pY-X3wdRDDUobI:b-ol0t8l) ([raw file](https://github.com/dgraph-io/badger/blob/6ade7de36a488ff5517f4ba4f9a6ff99ec7f16e0/table/mmap_windows.go#L14)):*
> ```Go
> 		return nil, err
> 	}
> 	defer syscall.CloseHandle(handler)
> ```

No need for handler anymore?

---

*[table/mmap_windows.go, line 21 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/62#-KnMWFmWNbBxrSVeK1fy:-KnMWFmWNbBxrSVeK1fz:b4qym9h) ([raw file](https://github.com/dgraph-io/badger/blob/6ade7de36a488ff5517f4ba4f9a6ff99ec7f16e0/table/mmap_windows.go#L21)):*
> ```Go
> 	}
> 
> 	data := (*[1 << 30]byte)(unsafe.Pointer(mapData))[:size]
> ```

Why use a fixed size array? Even for small tables, this would always use 1GB array? Shouldn't this be a slice?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/62#-:-KnMWnx2oxTFx49MJd2M:b-kyfyos)*

**Comment by Motakjuq** [comment]
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[table/mmap_windows.go, line 10 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/62#-KnMW4GO-O15sw5DcO5c:-KnQhsJ9WLPh2_HDspXp:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/6ade7de36a488ff5517f4ba4f9a6ff99ec7f16e0/table/mmap_windows.go#L10)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Over 100 chars. Split into 2 lines.
</blockquote></details>

Done.

---

*[table/mmap_windows.go, line 14 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/62#-KnMW9pY-X3wdRDDUobH:-KnQhuL8lLLp8oGrzgNB:b-txc5ui) ([raw file](https://github.com/dgraph-io/badger/blob/6ade7de36a488ff5517f4ba4f9a6ff99ec7f16e0/table/mmap_windows.go#L14)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

No need for handler anymore?
</blockquote></details>

it's not needed outside of this function

---

*[table/mmap_windows.go, line 21 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/62#-KnMWFmWNbBxrSVeK1fy:-KnQi85ZQEmXMupIEAKt:bk1w2jy) ([raw file](https://github.com/dgraph-io/badger/blob/6ade7de36a488ff5517f4ba4f9a6ff99ec7f16e0/table/mmap_windows.go#L21)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Why use a fixed size array? Even for small tables, this would always use 1GB array? Shouldn't this be a slice?
</blockquote></details>

1- is the max possible dimension of the array. 2- It doesn't reallocate new memory, it uses only the amount of memory provided by `MapViewOfFile`. 3- `data` is a slice, the array is converted to a slice by `[:size]`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/62)*

File: badger/value.go [hunk 1]
```diff
@@ -462,7 +461,7 @@ type valueLog struct {
 }
 
 func (l *valueLog) fpath(fid uint16) string {
-	return fmt.Sprintf("%s/%06d.vlog", l.dirPath, fid)
+	return fmt.Sprintf("%s%s%06d.vlog", l.dirPath, string(os.PathSeparator), fid)
```
Comment (icholy) [inline badger/value.go]: You could use [filepath.Join](https://golang.org/pkg/path/filepath/#Join) here.

**Comment by manishrjain** [comment]
:lgtm: Thanks a lot for the change.

---

Reviewed 9 of 12 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/62#-:-KncUGAO_SsJ2kIf5uT3:b941syq)*

**Pull request merged**